### PR TITLE
Fix recursion depth issue for bigger clusters

### PR DESCRIPTION
**Pull request by anmolbabu**
tendrl-bug-id: Tendrl/performance-monitoring#178
Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

### CODE REVIEW

File: tendrl/performance_monitoring/sds/ceph/ceph_plugin.py [hunk 1]
```diff
@@ -162,34 +163,54 @@ def get_nw_node_interfaces(self, node_id, nw_type, cluster_id):
...
 
-    def get_most_used_pools(self, cluster_det):
+    def get_cluster_pool_ids(self, cluster_id):
+        pools = []
+        try:
+            etcd_pools = NS._int.client.read(
+                '/clusters/%s/Pools/' % cluster_id
+            )
+            for etcd_pool in etcd_pools.leaves:
+                pool_key_contents = etcd_pool.key.split('/')
```
Comment (shtripat) [inline tendrl/performance_monitoring/sds/ceph/ceph_plugin.py]: Can you add a sample here?

File: tendrl/performance_monitoring/sds/ceph/ceph_plugin.py [hunk 2]
```diff
@@ -225,59 +246,98 @@ def get_pool_status_wise_counts(self, cluster_id, cluster_det):
...
+            try:
+                etcd_pool_rbds = NS._int.client.read(
+                    '/clusters/%s/Pools/%s/Rbds' % (
+                        cluster_id,
+                        pool_id
+                    )
+                )
+                for etcd_pool_rbd in etcd_pool_rbds.leaves:
+                    etcd_pool_rbd_key_contents = etcd_pool_rbd.key.split('/')
+                    if len(etcd_pool_rbd_key_contents) == 7:
```
Comment (shtripat) [inline tendrl/performance_monitoring/sds/ceph/ceph_plugin.py]: Same as above

Comment (r0h4n) [inline tendrl/performance_monitoring/sds/ceph/ceph_plugin.py]: what is this comparison with 7?

File: tendrl/performance_monitoring/sds/glusterfs/glusterfs_plugin.py [hunk 3]
```diff
@@ -136,8 +137,40 @@ def get_brick_status_wise_counts(self, volumes_det, cluster_id):
...
+        volume_ids = []
+        try:
+            etcd_volume_ids = NS._int.client.read(
+                '/clusters/%s/Volumes' % cluster_id
+            )
+        except EtcdKeyNotFound:
+            return volume_ids
+        for etcd_volume in etcd_volume_ids.leaves:
+            etcd_volume_contents = etcd_volume.key.split('/')
+            if len(etcd_volume_contents) == 5:
```
Comment (shtripat) [inline tendrl/performance_monitoring/sds/glusterfs/glusterfs_plugin.py]: As above

Comment (r0h4n) [inline tendrl/performance_monitoring/sds/glusterfs/glusterfs_plugin.py]: please explain this comparison with 5

File: tendrl/performance_monitoring/sds/ceph/ceph_plugin.py [hunk 4]
```diff
@@ -162,34 +163,54 @@ def get_nw_node_interfaces(self, node_id, nw_type, cluster_id):
...
-    def get_most_used_pools(self, cluster_det):
+    def get_cluster_pool_ids(self, cluster_id):
+        pools = []
+        try:
+            etcd_pools = NS._int.client.read(
+                '/clusters/%s/Pools/' % cluster_id
+            )
+            for etcd_pool in etcd_pools.leaves:
+                pool_key_contents = etcd_pool.key.split('/')
+                if len(pool_key_contents) == 5:
```
Comment (r0h4n) [inline tendrl/performance_monitoring/sds/ceph/ceph_plugin.py]: comparing with 5? Is this to check if the /Pools is non-empty?

File: tendrl/performance_monitoring/sds/__init__.py [hunk 5]
```diff
@@ -311,12 +309,12 @@ def __init__(self):
         self.supported_sds = []
         self.load_sds_plugins()
 
-    def get_cluster_summary(self, cluster_id, cluster_det):
-        sds_name = cluster_det.get('TendrlContext', {}).get('sds_name')
+    def get_cluster_summary(self, cluster_id, cluster_name):
+        sds_name = central_store_util.get_cluster_sds_type(cluster_id)
```
Comment (r0h4n) [inline tendrl/performance_monitoring/sds/__init__.py]: change to central_store_util.get_cluster_sds_name

File: tendrl/performance_monitoring/aggregator/cluster_summary.py [hunk 6]
```diff
@@ -94,14 +101,14 @@ def parse_cluster(self, cluster_id, cluster_det):
                 'percent_used': float(percent_used)
             },
             iops=str(self.get_cluster_iops(cluster_id)),
-            hosts_count=self.parse_host_count(cluster_det.get('nodes', {})),
-            sds_type=cluster_det.get('TendrlContext', {}).get('sds_name'),
+            hosts_count=self.parse_host_count(cluster_id),
+            sds_type=central_store_util.get_cluster_sds_type(cluster_id),
```
Comment (r0h4n) [inline tendrl/performance_monitoring/aggregator/cluster_summary.py]: sds_type or sds_name?

File: tendrl/performance_monitoring/aggregator/cluster_summary.py [hunk 7]
```diff
@@ -18,15 +18,19 @@ def __init__(self):
...
             'total': 0,
             'down': 0,
             'crit_alert_count': 0,
             'warn_alert_count': 0
         }
-        for node_id, node_det in cluster_nodes.iteritems():
-            status = node_det.get('NodeContext', {}).get('status')
+        cluster_nodes = central_store_util.get_cluster_node_ids(cluster_id)
+        for node_id in cluster_nodes:
+            node_context = central_store_util.read(
```
Comment (r0h4n) [inline tendrl/performance_monitoring/aggregator/cluster_summary.py]: no expect etcd.EtcdKeyNotFound? what if /clusters/%s/nodes/%sNodeContext doesnt exist?

File: tendrl/performance_monitoring/aggregator/cluster_summary.py [hunk 8]
```diff
@@ -18,15 +18,19 @@ def __init__(self):
...
+                '/clusters/%s/nodes/%s/NodeContext'
```
Comment (r0h4n) [inline tendrl/performance_monitoring/aggregator/cluster_summary.py]: no string format for "/clusters/%s/nodes/%s/NodeContext" ? I think you need cluster id and node id here

File: tendrl/performance_monitoring/aggregator/cluster_summary.py (hunk 6 above)
Comment (anmolbabu) [inline tendrl/performance_monitoring/aggregator/cluster_summary.py]: Done

File: tendrl/performance_monitoring/aggregator/cluster_summary.py (hunk 7 above)
Comment (anmolbabu) [inline tendrl/performance_monitoring/aggregator/cluster_summary.py]: Done

File: tendrl/performance_monitoring/sds/__init__.py (hunk 5 above)
Comment (anmolbabu) [inline tendrl/performance_monitoring/sds/__init__.py]: Done

File: tendrl/performance_monitoring/utils/central_store_util.py [hunk 9]
```diff
@@ -391,10 +397,29 @@ def get_node_cluster_id(node_id):
...
+
 def get_node_names_in_cluster(cluster_id):
     ret_val = []
     nodes = read('/clusters/%s/nodes' % cluster_id)
     for node_id, node_det in nodes.iteritems():
         ret_val.append(node_det.get('NodeContext', {}).get('fqdn'))
     return ret_val
 
+
+def get_cluster_sds_type(self, cluster_id):
```
Comment (anmolbabu) [inline tendrl/performance_monitoring/utils/central_store_util.py]: Done

File: tendrl/performance_monitoring/sds/ceph/ceph_plugin.py (hunk 4 above)
Comment (anmolbabu) [inline tendrl/performance_monitoring/sds/ceph/ceph_plugin.py]: Yes also some times I have seen things like /clusters/GlobalDetails or some such without cluster id so I did it this way

File: tendrl/performance_monitoring/sds/ceph/ceph_plugin.py (hunk 2 above)
Comment (anmolbabu) [inline tendrl/performance_monitoring/sds/ceph/ceph_plugin.py]: Yes also some times I have seen things like /clusters/GlobalDetails or some such without cluster id so I did it this way

File: tendrl/performance_monitoring/sds/glusterfs/glusterfs_plugin.py (hunk 3 above)
Comment (anmolbabu) [inline tendrl/performance_monitoring/sds/glusterfs/glusterfs_plugin.py]: Yes also some times I have seen things like /clusters/GlobalDetails or some such without cluster id so I did it this way

File: tendrl/performance_monitoring/sds/ceph/ceph_plugin.py (hunk 1 above)
Comment (anmolbabu) [inline tendrl/performance_monitoring/sds/ceph/ceph_plugin.py]: Done

File: tendrl/performance_monitoring/sds/ceph/ceph_plugin.py (hunk 2 above)
Comment (anmolbabu) [inline tendrl/performance_monitoring/sds/ceph/ceph_plugin.py]: Done

File: tendrl/performance_monitoring/sds/glusterfs/glusterfs_plugin.py (hunk 3 above)
Comment (anmolbabu) [inline tendrl/performance_monitoring/sds/glusterfs/glusterfs_plugin.py]: Done

**Comment by r0h4n** [comment]
Verified?

**Comment by anmolbabu** [comment]
Verified

**Comment by anmolbabu** [comment]
@r0h4n @nthomas-redhat @shtripat Please review

File: tendrl/performance_monitoring/aggregator/cluster_summary.py (hunk 8 above)
Comment (anmolbabu) [inline tendrl/performance_monitoring/aggregator/cluster_summary.py]: Done

**Pull request merged**